diff --git a/vendor/mlir/SimpleOperationInfo.h b/vendor/mlir/SimpleOperationInfo.h
index c2c50a0..a2e6b04 100644
--- a/vendor/mlir/SimpleOperationInfo.h
+++ b/vendor/mlir/SimpleOperationInfo.h
@@ -25,8 +25,7 @@ struct SimpleOperationInfo : public llvm::DenseMapInfo<Operation *> {
     auto *rhs = const_cast<Operation *>(rhsC);
     if (lhs == rhs)
       return true;
-    if (lhs == getTombstoneKey() || lhs == getEmptyKey() ||
-        rhs == getTombstoneKey() || rhs == getEmptyKey())
+    if (!lhs || !rhs)
       return false;
     return OperationEquivalence::isEquivalentTo(
         const_cast<Operation *>(lhsC), const_cast<Operation *>(rhsC),
